### PR TITLE
perf_watch_point.py:Testcase for perf watchpoint

### PR DESCRIPTION
**Pull request by shirishaganta**
Adding testcase for perf watchpoint data corruption
bug

Signed-off-by: Shirisha Ganta <shiganta@in.ibm.com>

### CODE REVIEW

File: perf/perf_watch_point.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import platform
+import shutil
+from avocado import Test
+from avocado import main
+from avocado import skipUnless
+from avocado.utils import cpu, build, distro, process
+from avocado.utils.software_manager import SoftwareManager
+
+CPU_ARCH = cpu.get_cpu_arch().lower()
+IS_POWER8 = 'power8' in CPU_ARCH
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py]: should  be like 

`IS_POWER8 = 'power8' in cpu.get_cpu_arch().lower()` as  minimize use of extra variable where ever possible

File: perf/perf_watch_point.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class PerfWatchPoint(Test):
+
+    def setUp(self):
+        '''
+        Install the basic packages to support perf
+        '''
+
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py]: this white space can be removed

File: perf/perf_watch_point.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        shutil.copyfile(self.get_data('swap.S'),
+                        os.path.join(self.teststmpdir, 'swap.S'))
+        shutil.copyfile(self.get_data('wptest.c'),
+                        os.path.join(self.teststmpdir, 'wptest.c'))
+        shutil.copyfile(self.get_data('Makefile'),
+                        os.path.join(self.teststmpdir, 'Makefile'))
+
+        build.make(self.teststmpdir)
+        os.chdir(self.teststmpdir)
+        process.run("insmod wptest.ko")
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py]: can we need to have a check if insmode  failed we need to fail test case

File: perf/perf_watch_point.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    def run_cmd(self):
+        i = 1
+        while i <= 4:
+            lst = process.run("cat /dev/wptest")
+            s = list(lst.stdout.strip().split(' '))
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py]: one question is this test case tested against upstream avocado means python3

File: perf/perf_watch_point.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            for result in s:
+                if result not in val:
+                    self.fail("wptest values are not correct")
+            i = i+1
+
+    @skipUnless(IS_POWER8, 'Supported only on Power8')
+    def test_watch_point_check(self):
+        if os.path.exists('/dev/wptest'):
+            self.run_cmd()
+            tmp = process.run("grep arg1 /proc/kallsyms")
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py]: please use avocado genio utility for file processing or python way

File: perf/perf_watch_point.py.data/wptest.c [hunk 6]
```diff
@@ -0,0 +1,56 @@
+#include <linux/module.h>
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py.data/wptest.c]: can we have author signature also in these files

File: perf/perf_watch_point.py (hunk 1 above)
Comment (shirishaganta) [reply perf/perf_watch_point.py]: yeah sure

File: perf/perf_watch_point.py (hunk 2 above)
Comment (shirishaganta) [reply perf/perf_watch_point.py]: ok

File: perf/perf_watch_point.py (hunk 4 above)
Comment (shirishaganta) [reply perf/perf_watch_point.py]: yes praveen i tested
JOB ID     : 4167ca981c3b61adb6e3edc03c9fe912b344a274
JOB LOG    : /root/avocado/job-results/job-2019-09-19T00.08-4167ca9/job.log
 (1/1) perf_watch_point.py:PerfWatchPoint.test_watch_point_check: PASS (3.66 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 4.05 s

File: perf/perf_watch_point.py (hunk 5 above)
Comment (shirishaganta) [reply perf/perf_watch_point.py]: ok

File: perf/perf_watch_point.py.data/wptest.c (hunk 6 above)
Comment (shirishaganta) [reply perf/perf_watch_point.py.data/wptest.c]: ok

Comment (shirishaganta) [reply perf/perf_watch_point.py.data/wptest.c]: yeah sure

File: perf/perf_watch_point.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        elif self.distro_name in ['rhel', 'SuSE', 'fedora', 'centos']:
+            deps.extend(['perf', 'kernel-devel'])
+        else:
+            self.cancel("Install the package for perf supported \
+                         by %s" % detected_distro.name)
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+        shutil.copy
+        archive.extract(self.get_data("wptest-master.tar.gz"), self.workdir)
```
[the remark this answers is not in the document]
Comment (shirishaganta) [reply perf/perf_watch_point.py]: @PraveenPenguin as i discussed with you actually facing an issue while doing make.so its better to keep tarfile itself praveen

File: perf/perf_watch_point.py (hunk 4 above)
Comment (PraveenPenguin) [reply perf/perf_watch_point.py]: please have this to work on python3 list(lst.stdout.decode('utf-8').strip().split(' '))

**Comment by shirishaganta** [comment]
@PraveenPenguin i tested for python3 also praveen it is working fine

File: perf/perf_watch_point.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        while i <= 4:
+            lst = process.run("cat /dev/wptest")
+            s = list(lst.stdout.decode('utf-8').strip().split(' '))
+            val = ['10', '20']
+            for result in s:
+                if result not in val:
+                    self.fail("wptest values are not correct")
+            i = i+1
+
+    @skipUnless(IS_POWER8, 'Supported only on Power8')
```
Comment (PraveenPenguin) [inline perf/perf_watch_point.py]: move this to setup section as setup also not required on this condition

Comment (shirishaganta) [reply perf/perf_watch_point.py]: yeah ok

**Pull request merged**